File: skills/render-create/templates/background-tasks/ts/worker.ts
/**
 * Background Worker
 *
 * Long-running background process for async tasks.
 * Runs continuously and processes work items.
 */

import "dotenv/config";

const POLL_INTERVAL = 5000; // 5 seconds

interface WorkItem {
  id: string;
  data: unknown;
}

/**
 * Process a single work item
 */
async function processWorkItem(item: WorkItem): Promise<void> {
  console.log(`Processing work item: ${item.id}`);

  // TODO: Add your processing logic here
  // Example: fetch from queue, process data, update database

  await new Promise((resolve) => setTimeout(resolve, 100)); // Simulate work

  console.log(`Completed work item: ${item.id}`);
}

/**
 * Fetch pending work items
 */
async function fetchWorkItems(): Promise<WorkItem[]> {
  // TODO: Implement your work item fetching logic
  // Example: poll a database, queue, or API

  // Placeholder: return empty array (no work)
  return [];
}

/**
 * Main worker loop
 */
async function runWorker(): Promise<void> {
  console.log("Worker started");
  console.log(`Poll interval: ${POLL_INTERVAL}ms`);

  while (true) {
    try {
      const items = await fetchWorkItems();

      for (const item of items) {
        await processWorkItem(item);
      }

      if (items.length === 0) {
        // No work, wait before polling again
        await new Promise((resolve) => setTimeout(resolve, POLL_INTERVAL));
      }
    } catch (error) {
      console.error("Worker error:", error);
      // Wait before retrying on error
      await new Promise((resolve) => setTimeout(resolve, POLL_INTERVAL));
    }
  }
}

// Handle graceful shutdown
process.on("SIGTERM", () => {
  console.log("SIGTERM received, shutting down gracefully...");
  process.exit(0);
});

process.on("SIGINT", () => {
  console.log("SIGINT received, shutting down gracefully...");
  process.exit(0);
});

// Start the worker
runWorker().catch((error) => {
  console.error("Fatal worker error:", error);
  process.exit(1);
});

File: skills/render-create/templates/background-tasks/py/worker.py
"""
Background Worker

Long-running background process for async tasks.
Runs continuously and processes work items.
"""

import os
import signal
import sys
import time
from dataclasses import dataclass
from typing import Any

from dotenv import load_dotenv

load_dotenv()

POLL_INTERVAL = 5  # seconds


@dataclass
class WorkItem:
    id: str
    data: Any


# Graceful shutdown flag
shutdown_requested = False


def signal_handler(signum, frame):
    """Handle shutdown signals gracefully."""
    global shutdown_requested
    print(f"Received signal {signum}, shutting down gracefully...")
    shutdown_requested = True


def process_work_item(item: WorkItem) -> None:
    """Process a single work item."""
    print(f"Processing work item: {item.id}")

    # TODO: Add your processing logic here
    # Example: fetch from queue, process data, update database

    time.sleep(0.1)  # Simulate work

    print(f"Completed work item: {item.id}")


def fetch_work_items() -> list[WorkItem]:
    """Fetch pending work items."""
    # TODO: Implement your work item fetching logic
    # Example: poll a database, queue, or API

    # Placeholder: return empty list (no work)
    return []


def run_worker() -> None:
    """Main worker loop."""
    print("Worker started")
    print(f"Poll interval: {POLL_INTERVAL}s")

    while not shutdown_requested:
        try:
            items = fetch_work_items()

            for item in items:
                if shutdown_requested:
                    break
                process_work_item(item)

            if not items:
                # No work, wait before polling again
                time.sleep(POLL_INTERVAL)

        except Exception as e:
            print(f"Worker error: {e}")
            # Wait before retrying on error
            time.sleep(POLL_INTERVAL)

    print("Worker shutdown complete")


if __name__ == "__main__":
    # Register signal handlers
    signal.signal(signal.SIGTERM, signal_handler)
    signal.signal(signal.SIGINT, signal_handler)

    try:
        run_worker()
    except Exception as e:
        print(f"Fatal worker error: {e}")
        sys.exit(1)
